### PR TITLE
SOLR-16682: transfer mlt query for component right

### PR DESCRIPTION
**Pull request by mkhludnev**
https://issues.apache.org/jira/browse/SOLR-788
I think this is a somewhat reliable solution for the problem with distributed mlt component. Reminder, now it naively relies on parsing `booleanQuery.toString()` that's unacceptable.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/MoreLikeThisHandler.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/handler/MoreLikeThisHandler.java:427]: LOL.... so it can help make it less ugly to name the parameter outputTerms or something like that.  I've also seen a strategy where the return value is a Pair or something like that.  There is such a class in SolrJ.

File: solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java:131]: We strive to match the Google Java Style guide, and that means using "Mlt" and not "MLT" inside names.  But maybe MLT is so pervasive capitalized this way in the code already (before striving for this) that you are trying to be consistent?  Any way; do as you wish.

**Review by dsmiley — COMMENTED** [review]
I just did a *super* quick review looking at some things that stood out to me.  Really Alessandro should review this.

File: solr/core/src/java/org/apache/solr/handler/MoreLikeThisHandler.java
Comment (mkhludnev) [reply solr/core/src/java/org/apache/solr/handler/MoreLikeThisHandler.java:427]: Got it. This collection also let to bypass terms extraction. I'm not sure it allows to save a lot of computation, I prefer to keep it lazy. I tries to toss code a little to improve. The root cause of this ugliness is lack of proper public method in Lucene. It's worth to decouple terms extraction, field assignment and query factoring. It will let to boost clauses without boolean query ripping.

File: solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java
Comment (mkhludnev) [reply solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java:169]: Thanks for suggestion, but I don't think it often matters much for field name.

File: solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java
Comment (mkhludnev) [reply solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java:131]: We have a hundred of occurrences of `MLT`, it might be great idea, but out of scope.

Comment (mkhludnev) [reply solr/core/src/java/org/apache/solr/handler/component/MoreLikeThisComponent.java]: ok. I put it as `mltViaQueryParams`

**Review by alessandrobenedetti — APPROVED** [review]
I have done a deep review and this pull request seems reasonable to me.
It's true that it mixes a nice refactor of the interesting terms extraction and communication with the shards (effectively solving two problems at once).
But I believe it's fine.
I left some minor comments that we should probably address, but not blockers

File: solr/core/src/test/org/apache/solr/handler/component/DistributedMLTComponentTest.java
Comment (mkhludnev) [reply solr/core/src/test/org/apache/solr/handler/component/DistributedMLTComponentTest.java:415]: It boils down to org.apache.solr.BaseDistributedSearchTestCase#query(boolean, org.apache.solr.common.params.SolrParams) which compares distributed and standalone responses. // ..so, if both of them returns nothing, it's perfectly fine ;)

**Comment by mkhludnev** [comment]
Thanks a lot @alessandrobenedetti

**Comment by alessandrobenedetti** [comment]
You are welcome! I saw you changed the anonymous 'id' to 'docNum'.
It's probably fine, but it didn't look intuitive to me as a name so I checked in the Solr code base, and it seems 'docId' to be more present and widespread used?
You know it's a super minor thing, I have a soft preference for using 'docId' but up to you!
And thanks for the nice work and refactoring, the MoreLikeThis needed some love, I really appreciate it!

**Comment by mkhludnev** [comment]
I think more often about Lucene `docNums`. Ok turned to id. Do you think it's worth to bother with open some MLT methods in Lucene to allow to grab interesting terms without ripping boolean query? Or it's worth to think about generic query serialization framework?

**Comment by alessandrobenedetti** [comment]
Back in the days, in 2017 when I was not a committer yet, I did a massive refactor of the Lucene More Like This code (https://www.slideshare.net/SeaseLtd/how-the-lucene-more-like-this-works, https://issues.apache.org/jira/browse/LUCENE-8326) and also presented it at various conf.
But I didn't get much traction and the work basically got wasted (a lot of tests and code cleaned up)
Long time has passed so it's very likely I would do it differently nowadays but I don't have the time (or fundings?) to do it again right now.
So in general I am in total favor of refactoring the Lucene side of things, happy to review it, I can't guarantee when, but I'll do my best!

**Comment by mkhludnev** [comment]
I'll push it tomorrow. Please chime it.

**Pull request merged**
